crates/autopilot/src/database/onchain_order_events/ethflow_events.rs

```diff
@@ -0,0 +1,208 @@
+use crate::database::events::meta_to_event_index;
+use anyhow::{anyhow, Context, Result};
+use contracts::cowswap_onchain_orders::{
+    event_data::OrderPlacement as ContractOrderPlacement, Event as ContractEvent,
+};
+use database::{
+    ethflow_orders::EthOrderPlacement, events::EventIndex,
+    onchain_broadcasted_orders::OnchainOrderPlacement, orders::Order, PgTransaction,
+};
+use ethcontract::Event as EthContractEvent;
+use std::{collections::HashMap, convert::TryInto};
+use unzip_n::unzip_n;
+
+use super::{CustomOnchainOrderParsing, CustomParsedOnchaninData};
+
+unzip_n!(pub 3);
+
+pub struct EthFlowOnchainOrderParser {}
+
+#[async_trait::async_trait]
+impl<'a> CustomOnchainOrderParsing<'a, u32, EthOrderPlacement> for EthFlowOnchainOrderParser {
+    fn parse_custom_event_data(
+        &self,
+        contract_events: &[EthContractEvent<ContractEvent>],
+    ) -> Result<Vec<(EventIndex, CustomParsedOnchaninData<u32>)>> {
+        contract_events
+            .iter()
+            .filter_map(|EthContractEvent { data, meta }| {
+                let meta = match meta {
+                    Some(meta) => meta,
+                    None => return Some(Err(anyhow!("event without metadata"))),
+                };
+                let ContractEvent::OrderPlacement(event) = data;
+                let quote_and_valid_to_touple = convert_to_quote_id_and_user_valid_to(event);
+                match quote_and_valid_to_touple {
+                    Ok(touple) => Some(Ok((
+                        meta_to_event_index(meta),
+                        CustomParsedOnchaninData {
+                            quote_id: touple.0,
+                            additional_data: Some(touple.1),
+                        },
+                    ))),
+                    Err(err) => {
+                        tracing::debug!(
+                            "Error while converting quote id and user valid to: {:?}",
+                            err
+                        );
+                        None
+                    }
```

Comment:
The error handling seems to be inconsistent. A few lines up missing metadata leads to an error being returned. Here the error only gets logged but it doesn't end up in the result vector.
Should the return type of this function maybe rather be `Vec<Result<_, _>>` instead of `Result<Vec<_>>` and have one entry in the result vector for every processed `EthContractEvent`?
Collecting `Iterator<Item = Result<T>>` into `Result<Vec<T>>` is kind of special so maybe this happens here unintentionally [[0](https://doc.rust-lang.org/std/iter/trait.Iterator.html#examples-28) (last example in the list)]?

Reply:
Yes!
I am not sure about it, but my thinking was:
The first error should never happen. Event data should always be there and consistent. Hence, an error should be propagated. ON the other hand, the data send as payload is not required to have the correct form all the time. Here, we just want to filter out the orders as invalid and not consider them.

We talked about it also here: https://github.com/cowprotocol/services/pull/536#discussion_r967677569. I would love to know how you would deal with it.

Reply:
Thanks for the pointer. So it's a very deliberate decision.
To me personally it feels a little weird to have a parsing function where the caller can't easily tell which items failed to parse (because they were dropped). I would expect a simple parsing function to be less opinionated and simply return different kind of parsing errors for each item. Then the caller can match on those errors and decide whether they are critical or to be ignored.
However, since this is a matter of taste and the expected behavior of the function is well documented in the trait (which I didn't see before) I think it's alright after all.

Reply:
You gotta point over here.
Generally speaking, I think 
- one should return the error messages to the calling function, if one wants to do the same action on the same errors for all custom_parsers and "steer it" more general.
- one should just leave the error handling to the custom_parsers, if one wants to grant maximum flexibility to the custom implementations. 

I can not really foresee what is more important and which option has less complexity. I think I would slightly prefer to keep it as it is. But I am really open to changing it. I am also curious about what others are thinking.

Reply:
My 2¢: I think the current behaviour makes sense:
- Return an error since we got unexpected data from the caller/node (we need metadata with the events, but it is missing)
- Don't return an error because someone manually executed a transaction with not enough data to the EthFLow contract - we skip this order because we don't understand it, but its business as usual.